rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+* permutation argument (existing, remove `n-ZK_ROWS` check) (elided)
+* lookup argument (existing) (elided)
+* runtime-table consistency
+  ```
+  RuntimeTableSelector * sum[i=1, X, RuntimeTable(i) * joint_combiner^i]
+   = sum[i=1, X, RuntimeTable(i) * joint_combiner^i]
+  ```
+* fixed and runtime table evaluation
+  ```
+  FixedAndRuntimeTable
+    = sum[i=0, W, FixedTable(i) * joint_combiner^i]
+      + sum[i=1, X, RuntimeTable(i) * joint_combiner^i]
+      + TableID * joint_combiner^max_joint_size
+  ```
+* lookup permutation argument
+  ```
+  LookupPermutation
+   * PermutationAggreg
+   * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+       + gamma + beta * Sigma(7))
+   * (SideLoadedCombined(0) + gamma + beta * Sigma(8))
+   * (SideLoadedCombined(1) + gamma + beta * Sigma(9))
+   * (SideLoadedCombined(2) + gamma + beta * Sigma(10))
+   * (SideLoadedCombined(3) + gamma + beta * Sigma(11))
+   * (RuntimeTable(1) + gamma + beta * Sigma(12))
+   =
+     LookupPermutation[prev]
+     * PermutationAggreg[prev]
+     * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+         + gamma + x * beta * shift[7])
+     * (SideLoadedCombined(0) + gamma + x * beta * shift[8])
+     * (SideLoadedCombined(1) + gamma + x * beta * shift[9])
+     * (SideLoadedCombined(2) + gamma + x * beta * shift[10])
+     * (SideLoadedCombined(3) + gamma + x * beta * shift[11])
+     * (RuntimeTable(1) + gamma + x * beta * shift[12])
+  ```
+* lookup permutation initializer
+  ```
+  Lagrange(0)
+    * (LookupPermutation * (0 + gamma + beta * zero_sigma)
+      - (0 + gamma + beta * shift[len(shift)-1]))
+    = 0
+  ```
+* lookup permutation finalizer
+  ```
+  Lagrange(n-ZK_ROWS) * (LookupPermutation - 1) = 0
+  ```
+
+This results in 5 new checks, and the removal of 1 old check.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This proposal increases the size of proofs, and increases the number of checks
+that the verifier must perform, and thus the cost of the recusive verifier.
+This also increases the complexity of the proof system.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### Why is this design the best in the space of possible designs?
+
+This design combines the 3 different 'modes' of random access memory, roughly
+approximating `ROM -> RAM` loading, `disk -> RAM` loading, and `RAM r/w`. This
+gives the maximum flexibility for circuit designers, and each primitive has
+uses for projects that we are persuing.
+
+This design has been refined over several iterations to reduce the number of
+components exposed in the proof and the amount of computation that the verifier
+requires. The largest remaining parts represent the data itself and the method
+for combining them, where simplifying assumptions have been made to minimize
+their impact while still satisfying the design constraints.
+
+### What other designs have been considered and what is the rationale for not choosing them?
```

Comment:
Do cairo / STARKs support some of this functionality?  If so, how does it compare?

Reply:
In STARKs they use a huge table that they call the execution trace where they load all needed witness values, but it can also store flags or other elements that may be needed to represent the relation. Then each column of the table is converted into polynomials and their constraints work for both secret witnesses and the values these auxiliary columns take.